classify.c

```diff
@@ -616,6 +616,103 @@ static void add_new_irq(char *path, struct irq_info *hint, GList *proc_interrupt
 		log(TO_CONSOLE, LOG_WARNING, "add_new_irq: Failed to add irq %d\n", irq);
 }
 
+static struct irq_info * build_one_dev_irq(const char *dirname, int irq)
+{
+	struct dirent *entry;
+	DIR *msidir;
+	int irqnum = 0;
+	struct irq_info hint;
+	char path[PATH_MAX];
+	char devpath[PATH_MAX];
+	struct irq_info *new = NULL;
+
+	sprintf(path, "%s/%s/msi_irqs", SYSPCI_DIR, dirname);
+	sprintf(devpath, "%s/%s", SYSPCI_DIR, dirname);	
+	hint.irq = irqnum;
+	hint.class = IRQ_OTHER;
+	
+	msidir = opendir(path);
+	if (msidir) {
+		do {
+			entry = readdir(msidir);
+			if (!entry)
+				break;
+			irqnum = strtol(entry->d_name, NULL, 10);
+			if (irqnum == irq) {
+				hint.type = IRQ_TYPE_MSIX;
+				break;
+			}
+		} while (entry != NULL);
+		closedir(msidir);
+	}else {
+		sprintf(path, "%s/%s/irq", SYSPCI_DIR, dirname);
+		if (process_one_line(path, get_int, &irqnum) < 0)
+			return NULL;
+		if (irqnum == irq)
+			hint.type = IRQ_TYPE_LEGACY;
+	}
+	if (irqnum == irq) {
+		add_new_irq(devpath, &hint, NULL);
+		new = get_irq_info(irq);
+	}
+	return new;
+}
+
+static struct irq_info * build_dev_irq(int irq)
+{
+	DIR *devdir;
+	struct dirent *entry;
+	struct irq_info *new_irq = NULL;
+
+	devdir = opendir(SYSPCI_DIR);
+	if (devdir) {
+		do {
+			entry = readdir(devdir);
+			if (!entry)
+				break;
+			new_irq = build_one_dev_irq(entry->d_name, irq);
+			if (new_irq)
+				break;
+		} while (entry != NULL);
+		closedir(devdir);
+	}
+	return new_irq;
+}
+
```

Comment:
And this is almost an exact duplicate of rebuild_irq_db.  Instead of cloning functions and suiting them to your needs, find a way to reuse that code

Reply:
ok